### ISSUE TITLE
NullPointerException in com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady

### ISSUE THREAD

**Issue by chrismcband** [GoogleChrome/android-browser-helper#109]
**Describe the bug**
We have recently upgraded from the google custom tabs client to android browser helper and notice a significant number of crashes, the highest being related to a NullPointerException from com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady

**To Reproduce**
1. Download Badoo Lite from play store using one of the mentioned devices in screenshot.
2. Launch the app

Errors are recorded when the app launches, play console has this stack trace from a Motorola Moto C Plus, but see also attached affected devices):

```
java.lang.NullPointerException: 
  at androidx.browser.trusted.TrustedWebActivityIntentBuilder.build (TrustedWebActivityIntentBuilder.java:220)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady (TwaLauncher.java:229)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.lambda$launchWhenSessionEstablished$3$TwaLauncher (TwaLauncher.java:217)
  at com.google.androidbrowserhelper.trusted.-$$Lambda$TwaLauncher$Pgq1RErTrlHE2hqdSd7zUQR6Yz4.run (lambda)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.onSplashImageTransferred (PwaWrapperSplashScreenStrategy.java:202)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.lambda$configureTwaBuilder$0$PwaWrapperSplashScreenStrategy (PwaWrapperSplashScreenStrategy.java:195)
  at com.google.androidbrowserhelper.trusted.splashscreens.-$$Lambda$PwaWrapperSplashScreenStrategy$w9o5iqMEy62ZZ2i_sMAmRiQ7IGw.onFinished (lambda)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:144)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:91)
  at android.os.AsyncTask.finish (AsyncTask.java:660)
  at android.os.AsyncTask.-wrap1 (AsyncTask.java)
  at android.os.AsyncTask$InternalHandler.handleMessage (AsyncTask.java:677)
  at android.os.Handler.dispatchMessage (Handler.java:110)
  at android.os.Looper.loop (Looper.java:203)
  at android.app.ActivityThread.main (ActivityThread.java:6251)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1075)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:936)
```

**Expected behavior**
App launches without the NullPointerException

**Did this ever used to work**
Similar error happened with google tabs client.

**Screenshots**
<img width="972" alt="Screenshot 2020-06-08 at 12 56 24" src="https://user-images.githubusercontent.com/196567/84032738-aca2f680-a98f-11ea-8860-e2e2fc83c120.png">
<img width="640" alt="Screenshot 2020-06-08 at 13 04 36" src="https://user-images.githubusercontent.com/196567/84032769-b7f62200-a98f-11ea-8bd7-d9f2fbadb39a.png">

**Code Snippets**
Splash screen strategy is being set with this code:

```
this.mSplashScreenStrategy = new PwaWrapperSplashScreenStrategy(this, this.mMetadata.splashImageDrawableId, this.getColorCompat(this.mMetadata.splashScreenBackgroundColorId), this.getSplashImageScaleType(), this.getSplashImageTransformationMatrix(), this.mMetadata.splashScreenFadeOutDurationMillis, this.mMetadata.fileProviderAuthority);
```

These all have values except for transformation matrix which is null.

**Smartphone (please complete the following information):**
 - Device: Motorola Moto C Plus
 - OS: Android 8.1
 - Browsers Installed: Chrome
 - Browser Versions: Don't know
 - android-browser-helper library version: 1.1.0

**Additional context**
Badoo Lite is released on the play store for low end devices as an alternative for the native android app.

**Comment by andreban** [comment]
@chrismcband This seems to be the issue fixed by #92. Can you check if this still reproducible when using `1.3.1-alpha01`?

**Comment by chrismcband** [comment]
I will have to get my hands on one of the phones that has this issue, I've got one ordered and will post the results here.

**Comment by andreban** [comment]
Cool - I couldn't download the app to my test device (Samsung A10 / SM-A105FN), which may be able to repro, as it's not available to download in the UK. Would you be able to share the APK with me? (maybe via e-mail, if that's better)

**Comment by chrismcband** [comment]
@andreban I can send you an APK, what's your email? Might want to obfuscate it a bit.

**Comment by andreban** [comment]
@chrismcband same as my Github username, but at google.com

**Comment by chrismcband** [comment]
@andreban I've shared a file by google drive. Attachment in email failed because it was blocked by security for content reasons.

**Comment by chrismcband** [comment]
@andreban I tried the APKs on a Motorola Moto C Plus, I managed to replicate the NullPointerException crash with the 1.1 version. It was very temperamental, occuring between 15% - 40% of runs.
For the 1.3.1 alpha library I didn't get it once and I tried about 60 times. However I did get several occasions where it was stuck on the splash screen, almost at the same rate, this never happened on the 1.1 version. I'll try to do more digging. Also I noticed that the url bar was showing up much more often in the 1.3.1 alpha library. I saw it maybe once or twice in 40 runs on 1.1 and around 13 times in 60 runs on 1.3.1.

**Comment by andreban** [comment]
Thanks for checking this. I'll try to get a similar device to repro - It seems the Moto C Plus comes in 1 GB and 2 GB RAM versions. Do you know which one you are testing on?

**Comment by chrismcband** [comment]
It's the 1 GB one.

**Comment by andreban** [comment]
Trying to repo on an older Moto E 2nd gen (also 1GB, but with Android 6), without success. I wonder if it's more device specific or some other state in the phone.

**Comment by chrismcband** [comment]
I have Android 7 on mine. I had to try it many times before I got it. But now I can usually get it at least once every 10 times. This is swiping away the activity each time of course.

**Comment by andreban** [comment]
Do you have other apps open at the same time?

**Comment by chrismcband** [comment]
No nothing else open.

**Comment by andreban** [comment]
Tried on an emulator - I couldn't repo in normal conditions, but managed to do it from adb, with the command below:
```
adb shell am start -n com.badoo.twa/.BadooLiteLauncherActivity && adb shell am force-stop com.android.chrome
```

@chrismcband I'm not familiar with the particular device, but from the adb repo it seems the system is killing Chrome during the startup. My question is: does this device have some kind of memory optimizer / battery optimizer enabled? If so, can you try disabling it if check if it's still reproducible?
Thanks for all the help!

**Comment by chrismcband** [comment]
Battery saver is off in android settings. Under Location settings it says Mode Battery saving though, but in Battery settings it has Battery saver Off / Never turn on automatically. Can't see any memory optimizer setting.

**Comment by andreban** [comment]
Checking out what may be happening, I think the issue may be on [how `CustomTabsClient` binds to the Chrome Service](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-browser-release/browser/src/main/java/androidx/browser/customtabs/CustomTabsClient.java#76). It binds using [`Context.BIND_WAIVE_PRIORITY`](https://developer.android.com/reference/android/content/Context#BIND_WAIVE_PRIORITY), which I think prevents Chrome from gaining increased priority, meaning it can be killed by the OS even if the foreground app is connected. Removing this may fix the issue.

It seems we'd need a change to `androidx.browser` itself to get the issue properly fixed, as one of the methods we'd need is package-only. I did create monkey-patch fix: https://github.com/andreban/android-browser-helper/commit/7be828a968dfdbe2c4015c8c8d289244d27734d1. Let me know if you want to try out this patch and confirm (I can generate the library to drop in the project if it helps).

I'll do more extensive testing next week.

**Comment by chrismcband** [comment]
@andreban yes sure, I can test your library. I've just been using gradle dependencies so if it's a different process let me know, thanks.

**Comment by andreban** [comment]
Thanks for this - I tried doing more repro today. A more realistic approach would be something like the below, as `force-stop` kills the app in more unnatural way - the kill command behaves should behave more like the OS killing background tasks to free up memory and will avoid killing a process that has high priority.
```
adb shell am start -n com.badoo.twa/.BadooLiteLauncherActivity && adb shell am kill com.android.chrome
```
But with this command, the issue doesn't happen :(. 

Anyway, I have added an AAR file [here](https://drive.google.com/file/d/1Ny1uWSgYFZDZsersbL6cvtnwA-UYzgUr/view?usp=sharing).

You can add this file to a `libs` subdirectory in project root, then modify the dependency section of `app/build.gradle` to something like the following. Not that the `.aar` extension was added on the first implementation line. We also need to include appcompat in order to import the base themes used. 

```gradle
dependencies {
    implementation fileTree(include: ['*.jar', '*.aar'], dir: 'libs')
    implementation 'androidx.appcompat:appcompat:1.1.0'
}
```

@chrismcband  Let me know if this changes anything for you, and really appreciate the help on this.

**Comment by chrismcband** [comment]
I tried including the .aar file and did what you said, but when I try and rebuild the project I see a bunch of errors like
`package androidx.browser.customtabs does not exist`
`package androidx.browser.trusted does not exist`
`cannot find symbol class TrustedWebActivityIntentBuilder`

I have tried with the libs directory in project root and at `app/libs`, but both have same result.

I'm using Android Studio 4.0

**Comment by andreban** [comment]
Got it - 

It seems the AAR will not trigger automatically including androidx.browser:

Can you also add `implementation 'androidx.browser:browser:1.2.0'` to the dependencies?

To confirm, yes, you should remove `implementation 'com.google.androidbrowserhelper:androidbrowserhelper:1.1.0'`;

```gradle
dependencies {
    implementation fileTree(include: ['*.jar', '*.aar'], dir: 'libs')
    implementation 'androidx.appcompat:appcompat:1.1.0'
   implementation 'androidx.browser:browser:1.2.0'
}
```

**Comment by chrismcband** [comment]
Ok now I see errors about:
`package com.google.androidbrowserhelper.trusted does not exist`
`cannot find symbol class TwaLauncher`

Basically anything with `com.google.androidbrowserhelper`

**Comment by chrismcband** [comment]
When I used this I could build:
```
    implementation fileTree(include: ['*.jar, *.aar'], dir: 'libs')
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'androidx.browser:browser:1.2.0'
    compile(name:'androidbrowserhelper-release', ext:'aar')
```

**Comment by chrismcband** [comment]
@andreban I think we have a winner, ran it 20 times, no issues, splash screen went away every time and didn't see a url bar once. I'll try it on my Samsung A10 and will check it with push notifications. So far it looks like best version yet.

**Comment by andreban** [comment]
Glad to hear this! Waiting to hear more on how tests on the Samsung A10 go.

**Comment by chrismcband** [comment]
20 perfect launches on the Samsung A10 also.

**Comment by andreban** [comment]
Good to have the likely root cause nailed. I'll work on figure out a way to temporarily fix this with a more elegant solution (I'm using reflection right now and that will give you warnings on the Play Store). We may need changes to `androidx.browser` for the definitive solution.

**Comment by chrismcband** [comment]
Thanks, please keep me updated here. We're keen to update and improve our crash rate.

**Comment by dalexanco** [comment]
Hi guys, we also have this issue on our latest TWA app, do you have a release planning ?
I understand that the bug is difficult to reproduce (I tried without success on my side), we have ~2% of errors from crashlytics due to this issue, not only on samsung A10. 

Since the PR is merged, when do you think the next version including this patch will be released ?

**Comment by andreban** [comment]
Hi @dalexanco, we did create a patch on https://github.com/andreban/android-browser-helper/commit/7be828a968dfdbe2c4015c8c8d289244d27734d1 that works around the issue but haven't created a PR, as it uses reflection and will trigger warnings on the Play Store. 

Instead, we're looking into getting a patch on `androidx.browser`, which is the library `android-browser-helper` depends on and is what is triggering the error.  The release process for `androidx.browser` is a bit more involved and we're getting more details on timelines.

If we're not able to get the patch there in a reasonable amount of time, we may go forward and create a PR with that patch on android-browser-helper (which has a much simpler release process), and remove it once the fix on `androidx.browser` is released.

I'll keep this post updated.

**Comment by dalexanco** [comment]
Totally clear, @andreban thank you for the explanation ❤️

**Comment by andreban** [comment]
Quick update here is that, if everything goes right, we should have an alpha version of AndroidX by the middle of next week and an alpha version of android-browser-helper by the end of the week. (again, that's what we're aiming for if nothing else gets int he way)

**Comment by andreban** [comment]
We have released `1.4.0-alpha01` yesterday with the fix for this issue included:

```gradle
    implementation 'com.google.androidbrowserhelper:androidbrowserhelper:1.4.0-alpha01'
```

**Comment by rockeynebhwani** [comment]
@andreban - I don't see 1.4.0-alpha01 on releases page - https://github.com/GoogleChrome/android-browser-helper/releases

**Comment by chrismcband** [comment]
Hi @andreban do you have an idea when the official 1.4.0 with this fix will be released?

**Comment by andreban** [comment]
This depends on an updated to `androidx.browser`. Let me check for the timelines on that.

**Comment by chrismcband** [comment]
@andreban is there any update on when the fix can be released?

**Comment by chrismcband** [comment]
@andreban any updates?

**Comment by chrismcband** [comment]
@andreban is the fix for this include in the 2.0.1 release?

**Comment by andreban** [comment]
[androidx.browser 1.3.0 RC1](https://developer.android.com/jetpack/androidx/releases/browser#version_130_2) was released yesterday. We're still waiting for it to become stable (sorry, the androidx release slower than we'd like.

If this is critical,`com.google.androidbrowserhelper:androidbrowserhelper:2.1.0-alpha01` contains the fix and should be stable enough.

**Comment by andreban** [comment]
`androidx.browser-1.3.0` has finally hit stable and we have released version `2.1.0` of `android-browser-helper`, which depends on the new `androidx.browser` version. This bug shouldn't happen anymore when using that. To use it, update the app's `build.gradle` as below:

```
implementation 'com.google.androidbrowserhelper:androidbrowserhelper:2.1.0'
```

Thanks for the patience for this issue to be resolved.
